### ISSUE TITLE
[RFC]elements schema: add support for custom schemas

### ISSUE THREAD

**Issue by vicb** [angular/angular#12045]
in `@NgCompoent`, schemas can currently only use the following values:
- `CUSTOM_ELEMENTS_SCHEMA` -> allow any `custom-tag` with any property,
- `NO_ERRORS_SCHEMA` -> allow any tag and any property

We need to implement a way to defined custom schemas, ie defining all polymer elements and their properties (and events).

@mprobst when would you be available so that we can discuss how to integrate with the security schema ?

**Comment by mprobst** [comment]
I'll be around tomorrow.

Victor Berchet <notifications@github.com> schrieb am Mo., 3. Okt. 2016,
10:13:

**Comment by tbosch** [comment]
If we do this right, our current `DomSchema` should not be special cased any more, right?

**Comment by vicb** [comment]
Roberto Fischer would be interested in this feature

**Comment by laco0416** [comment]
I think #12840 is a similar request in terms of the customizable sanitization.

**Comment by BaerMitUmlaut** [comment]
I accidentally created a duplicate of this, here is my suggestion to resolve this (copy & pasted from the other issue):

---

I'd expand the SchemaMetadata interface like this:

```ts
export interface SchemaMetadata {
  name: string;
  tokenValid?: (token: string) => boolean;
}
```

The functions would check if the given token (element or attribute name) is valid, it's optional for full backwards compatibility (though that's probably not even necessary). Alternatively this could be split into two functions (attributeValid & elementValid).

Then there could be a class or helper function that generates an object which simply checks if the given token is within a set of previously defined tokens.

This would also clean up existing code because you don't have to hardcode the logic of the schema everywhere it's used.

If you approve, I can implement this and PR, should be pretty simple.

**Comment by trotyl** [comment]
> If you approve, I can implement this and PR, should be pretty simple.

@BaerMitUmlaut How do you make the compiler to **execute** a specific function in metadata, without running other code? The main problem is this isn't happening at runtime.

**Comment by BaerMitUmlaut** [comment]
I'm not an Angular pro, but these are the two spots where I could find logic associated with `CUSTOM_ELEMENTS_SCHEMA` (it's less often used overall, thus it was easier to find, I'd guess these are also all the places where `NO_ERRORS_SCHEMA` gets handled).
https://github.com/angular/angular/blob/b0578061cec332397bbd12cadd2fddfca3fe257f/packages/core/src/render3/instructions/shared.ts#L933-L934
https://github.com/angular/angular/blob/d11b0c0c415d4c3584eb1e1c83bf80e9ee61dcf8/packages/compiler/src/schema/dom_element_schema_registry.ts#L295
They would have to be changed to call the `tokenValid` method instead. Unless there is some other magic happening somewhere else which I'm missing, that should be it, right?

**Comment by trotyl** [comment]
@BaerMitUmlaut The ivy one is not feature complete and currently WIP at https://github.com/angular/angular/pull/30301

Also, the schemas must be understood by language-service, thus cannot be runtime-only.

**Comment by Jack-Works** [comment]
```typescript
import { MyComponent } from './my-comp'
@NgModule({
  declarations: [AppComponent],
  imports: [BrowserModule],
  providers: [],
  bootstrap: [AppComponent],
  customElements: [MyCompoent]
})
export class AppModule {}
```

> my-comp.ts

```typescript
export interface MyComponent {
    customData: [string, number]
    onChange: (data: string) => void
}
export const MyComponent = { tag: 'my-comp' } as const
class MyComp extends HTMLElement {
    // .......
}
```

```html
<!-- The exported value ensure angular recognize it's tag name -->
<!-- The exported interface ensure angular compiler check all properties of `my-comp` component -->
<!-- They shared the same name "MyCompoent" so angular compiler can combine them together. -->
<!-- Now customData get the type [string, number], $event in "onChange" is type "string" -->
<my-comp [customData]="data" (onChange)="onChange($event)"></my-comp>
```

**Comment by RobM-ADP** [comment]
I don't think this should be considered low priority or frequency. We are probably all starting to see the shifting of design system component libraries being built using web components and this makes it very difficult for devs to work with them inside angular apps. Right now to use custom elements you must essentially turn off all schema validation. An ideal solution (for us anyway) would be to be able to build and import a custom schema that was generated by the Stencil compiler.

**Comment by DmitryEfimenko** [comment]
I totally agree with @RobM-ADP. Just recently we had to do quite an ugly workaround to avoid the `CUSTOM_ELEMENTS_SCHEMA` in our app where we'd introduce a million of directives that would target custom elements that come from a 3rd party lib just so that Angular "knows" about them. In a lot of cases we had to add @Inputs() to these directives so that Angular does not complain about these. The thing is, when doing that, in some cases it changes the behavior of the custom component, which is not desired at all. To summarize, it's a super ugly workaround, but it's better than using `CUSTOM_ELEMENTS_SCHEMA`. It'd be nice if there was an official way to handle this scenario.

**Comment by RobM-ADP** [comment]
Another thing worth noting is that creating a lot of directives or wrapper components as @DmitryEfimenko describes just to make the compiler happy introduces a lot of extra code in the final bundles that doesn't need to be there. One of the big advantages of building a design system as webcomponents is that code defining the components doesn't need to exist in the Angular app's code. So neither `CUSTOM_ELEMENTS_SCHEMA` nor custom wrappers/directives are a good solution to a problem that is going to happen more frequently over time.

**Comment by CSchulz** [comment]
Even after #32171 was merged there is no possibility to add some own custom elements.

Before Ivy it was possible to inject some own ElementSchemaRegistry via the COMPILER_OPTIONS to check the elements during a JIT build. This was usable for ng serve and ng test.

@alexzuza It seems that your workaround by overwriting the DomElementSchemaRegistry doesn't work anymore for "ng serve"
https://medium.com/angular-in-depth/angular-elementschemaregistry-for-dummies-83d54cd31478
https://twitter.com/yurzui/status/1224676392744431616

It works fine with "ng build" but not with "ng serve" nor "ng test". To be more specific it does "work" for the first build iteration but then it throws up all errors like before.

Finally I can say, we need this feature to add some external schema definition to have a template validation.

**Comment by MichaelBrimer** [comment]
We use web components and because of this we use CUSTOM_ELEMENTS_SCHEMA. 
This gives a lot of problems, one solution is to make a wrapper, which other has described, but a described this is not a good solution for several reasons.
We have tried different things, among these to overwriting the DomElementSchemaRegistry. 
But we need a solution, which works BOTH with the build and in the Angular Language Service.
It seems at this can only be fixed by the Angular Team, so please help us.

**Comment by waterplea** [comment]
@MichaelBrimer do you have a solution that at least works in AOT builds of modern Angular? If I don't care about language service :) Overwriting the registry is tricky in current gen ng serve :(

**Comment by MichaelBrimer** [comment]
@waterplea No we could not get it to work with AOT.
We have 4 big Angular projects today and needs this feature, but people are starting ask if we ever get it and if Angular is a good choice.
The solution with CUSTOM_ELEMENTS_SCHEMA is so bad, that it is impossible to defend for us who loves Angular.
Google has +2.600 Angular projects, but the question is whether they will fix this issue if they do not use web components in these projects.
The Angular Team has just made a major update to the language service, without fixing this issue :-(

**Comment by waterplea** [comment]
Found another case that this issue is messing with — host bindings. Since they are part of parent view you cannot even rely on `NO_ERRORS_SCHEMA` because it might not present in the parent module.

**Comment by sijakret** [comment]
The fact that this feature is missing __is really slowing down adoption of web components__ in enterprise applications where angular is very prominent.
Not implementing this would effectively be a hostile move against the adoption of web components in angular apps.
Why can't there be a simple way to extend the set of default built in html tag definitions?

**Comment by wSedlacek** [comment]
This would be nice for non DOM environments as well, ie NativeScript.
Looking through the code I noticed that `DomElementSchemaRegistry` is created in a few places rather than using some form of dependency injection.

It would be cool if this could be provided as part of the platform module as the DOM is very specific to the web platform.

**Comment by sijakret** [comment]
is there any reason to not just provide a way to customize the following checks?
https://github.com/angular/angular/blob/b5ab7aff433a67cddaa55e621d17b1a1b07b57c2/packages/compiler/src/schema/dom_element_schema_registry.ts#L294
https://github.com/angular/angular/blob/b5ab7aff433a67cddaa55e621d17b1a1b07b57c2/packages/compiler/src/schema/dom_element_schema_registry.ts#L315

if these were pluggable users could already significantly refine which tags to bail out on.
this would already be much better than just dropping everything when CUSTOM_ELEMENTS_SCHEMA is used

if i patch my local compiler with 3 lines i can get immediately it to do what i need!

**Comment by waterplea** [comment]
This sounds reasonable. If it was to be the choice, default checks must be exposed too so we can do something like:
```TypeScript
override hasProperty(tagName: string, propName: string, schemaMetas: SchemaMetadata[]): boolean { 
  return propName.includes('MY_CUSTOM_PREFIX') || DEFAULT_CHECK(tagName, propName, schemaMetas);
}
```

**Comment by sijakret** [comment]
Why not even allow plugging in a full hook?
```ts
override hasProperty(tagName: string, propName: string, schemaMetas: SchemaMetadata[]): boolean { 
  return config.hasPropertyOverride ?  config.hasPropertyOverride(tagName, propName, schemaMetas) : DEFAULT_CHECK(tagName, propName, schemaMetas);
}
```

**Comment by wSedlacek** [comment]
> Why not even allow plugging in a full hook?

As far as I have looked into, the issue is how to expose it this configuration.
`DomElementSchemaRegistry` is used in the AOT compiler as far as I understand so standard dependency injection doesn't seem like it would cover us there.

**Comment by sijakret** [comment]
> > Why not even allow plugging in a full hook?
> 
> As far as I have looked into, the issue is how to expose it this configuration.
> `DomElementSchemaRegistry` is used in the AOT compiler as far as I understand so standard dependency injection doesn't seem like it would cover us there.

Ok. Would some sort of pattern or similar for the config work?

**Comment by sijakret** [comment]
Hi,

I cooked up the following proposal:

https://github.com/sijakret/angular/commit/a5cc45e121d5ba0437461cab84b452a04ebcfd82?branch=a5cc45e121d5ba0437461cab84b452a04ebcfd82&diff=unified

it adds a new angular compiler option `angularCompilerOptions.webComponentPattern`

```js
{
  ...
  "angularCompilerOptions": {
    ...
    "webComponentPattern": "my-prefix-" // RegEx
  }
}
```

if it is not used there should be no change in behavior.
if set + `CUSTOM_ELEMENTS_SCHEMA` is used then
* only tags that match the pattern will be considered as web components.
* all tags that don't match the pattern are checked in the default way (html5 tag or angular component)

Example (config option set like so: `"webComponentPattern": "my-prefix-"`)

```ts
@Component({
  selector: 'component-exists',
  template: `<div style="padding: 20px">{{prop}}</div>`,
})
export class ComponentExists {
  @Input() prop:string = 'default'
}

@NgModule({
  declarations: [
    // ...
    ComponentExists
  ],
  //...
  schemas: [CUSTOM_ELEMENTS_SCHEMA]
})
export class AppModule { }

```
template
```html

<!-- A
regular angular component with a prop
=> works
-->
<component-exists [prop]="'prop was set'"></component-exists>

<!-- B
same component, but trying to set prop with wrong type
=> error even though CUSTOM_ELEMENTS_SCHEMA is configured
-->
<component-exists [prop]="1"></component-exists>


<!-- C
same component, but trying to set non-existing prop
=> error even though CUSTOM_ELEMENTS_SCHEMA is configured
-->
<component-exists [propOoops]="'prop was set'"></component-exists>

<!-- D
trying to use an angular component that does not exist
=> error,.. 
-->
<component-does-not-exist></component-does-not-exist>

<!-- E
matches pattern /my-prefix-/i
=> no checks at all
-->
<my-prefix-component></my-prefix-component>
```
This behavior would be a significant upgrade to the current one since we can have E but still get validation feedback for B, C and D.
Works well in a test project with both `ng serve` and `ng build`.

Looking for feedback from the team here - do you see issues with this approach?
In case this path can prepare a PR according to the guidelines with tests etc.

I am aware this is not the most comprehensive or elegant solution but it would be very low-risk/high-impact imo.

@wSedlacek @CSchulz @waterplea

**Comment by MichaelBrimer** [comment]
Thanks, - this would be a great improvement !!

**Comment by waterplea** [comment]
@sijakret it wouldn't help with HostBinding, because they belong to the parent view, so you will not be able to guarantee that `CUSTOM_ELEMENTS_SCHEMA` is present there. It would also not help in my case, where I want to be able to store some arbitrary values as properties on elements. Zone.js does that with `__zone_symbol_xxx` why can't we? There's a standard valid way to store string values as attributes with `data-` but sometimes I need to store other types.

**Comment by sijakret** [comment]
@waterplea can you provide an example?

**Comment by waterplea** [comment]
If you had some directive that processes value somehow and do this, it would cause an error, even if you add `CUSTOM_ELEMENTS_SCHEMA` to the directive module:
```ts
@Directive({ selector: 'clamp' })
export class ClampDirective {
  @Input() min = 0;
  @Input() max = Infinite;
  @Input() value = 0;

  @HostBinding('prop')
  get prop(): number {
    return Math.min(Math.max(this.value, this.min), this.max);
  }
}
```
```html
<component-exists clamp [value]="value"></component-exists>
```
It's a fairly crazy example, I know, but I just wanted to highlight that there's also `@HostBinding` which participates in bindings and it's a special case because even if you define it in one component of one module, it is actually executed in another view in another module and follows that module rules. Besides, you need to consider libraries. How would compiler parameters for the library be respected in app that uses this library? I believe there are runtime checks which would through error currently.

**Comment by sijakret** [comment]
I cannot get your example to do what i think it is inteded to do regardless of any schema-related issues, here it is alive: https://stackblitz.com/edit/angular-ivy-8mzrpk?file=src%2Fapp%2Fapp.module.ts

Funny thing is, it works for the web component (CUSTOM_ELEMENT_SCHEMA and WC are commented out in the stackblitz since the problem with the directiv seem unrelated to me)
```
<!-- works, clamps and sets clamped prop -->
<my-prefix-custom-element clamp name="with clamp directive" [value]="10"></my-prefix-custom-element>
<!-- breaks, value should be number -> with the update i am suggeting it will even typecheck the directive  -->
<my-prefix-custom-element clamp name="with clamp directive" [value]="'10'"></my-prefix-custom-element>
```

Regarding libraries: I am not sure about the way angular handles this, but the compiler probably compiles these things AOT and hence use the compiler settings at compile time of the lib for its code (at least regarding templates). Don't know this though.

@alxhub looking forward to your thoughts on my proposal https://github.com/angular/angular/issues/12045#issuecomment-928587907

**Comment by wSedlacek** [comment]
Having just a prefix won't solve the Schema for NativeScript.
Really for that, we need the ability to inject a new schema for that platform.

**Comment by sijakret** [comment]
Instead of a RegExp pattern the config load external schemas, for instance as described here: https://github.com/microsoft/vscode-custom-data (example: https://github.com/microsoft/vscode-custom-data/blob/main/samples/helloworld/html.html-data.json) and check against them.
These are already supported by vscode - gives you autocomplete etc. also in conjunction with angular tooling.

This change would be a bit more involved but in essence a regex vs. a dictionary pretty much won't change much about the way it hooks into the compiler. At this point i am looking for feedback whether i am on the right track in the first place ;)

**Comment by sijakret** [comment]
This issue has been around since 2016, is not hard to fix and would help EVERY angular project that uses web components.
To me it feels like it is strategically ignored to keep up lock in.

**Comment by Abadii** [comment]
Are there any updates regarding this issue?

**Comment by Kim-Andersen** [comment]
I hate to agree that this issue increasingly smells like a lock-in strategy.
Custom Elements is a standard, just like DIV and DIALOG elements, and is here to stay, which makes it weird that Angular is doing nothing is this regard. It feels like a punishment when "opting out" with the `CUSTOM_ELEMENTS_SCHEMA`.

Angular team; please at least share a reason, roadmap, thoughts, anything.

**Comment by danwulff** [comment]
Can confirm this is affecting our organization. We're migrating our Web design system components from Angular to https://lit.dev/ in order to support framework-less frontends. It feels like we're being punished for using Custom Elements even though that's a web standard.

**Comment by CSchulz** [comment]
We have used the old way to patch the JIT compiler. Because this isn't functional we have switched to patching the angular core, but that is no straight forward way to go.
I think it could be a better option to do such things on a predefined way with extended diagnostics or similar. For this I have opened a feature request #48117.

**Comment by elgreco247** [comment]
> If you had some directive that processes value somehow and do this, it would cause an error, even if you add `CUSTOM_ELEMENTS_SCHEMA` to the directive module:

@waterplea Isn't it the case, that you always should add the CUSTOM_ELEMENTS_SCHEMA to the module declaring the component which uses the custom element? And in the case of standalone components, to the component itself?

**Comment by waterplea** [comment]
> > If you had some directive that processes value somehow and do this, it would cause an error, even if you add `CUSTOM_ELEMENTS_SCHEMA` to the directive module:
> 
> @waterplea Isn't it the case, that you always should add the CUSTOM_ELEMENTS_SCHEMA to the module declaring the component which uses the custom element? And in the case of standalone components, to the component itself?

I guess, but that means if you distribute reusable components you cannot bundle required config with it and people will have to do it themselves, which would then affect their code, which it really shouldn't.

**Comment by break-stuff** [comment]
Is there any update with this. We currently use this plugin to generate VS Code and JetBrains autocomplete and docs, but it would be nice if we could also support type safety for our components.

- https://www.npmjs.com/package/custom-element-vs-code-integration
- https://www.npmjs.com/package/custom-element-jet-brains-integration

**Comment by olaf-k** [comment]
hey @pkozlowski-opensource! would you mind weighing in on this? it'd be nice to get the team's position on this.
can we hold our breath for a future solution to feed the compiler custom elements schemas? is there any alternative to creating wrappers/directive?

**Comment by eiswind** [comment]
Sad to see that there is no progress on this. I`d also like to throw in web-types, which could be a way to even validate custom elements. I am used to this from vaadin components, but I have seen that it's already in use be some other component libraries.
It's simply a json schema to describe your component api.

https://github.com/JetBrains/web-types

**Comment by trusktr** [comment]
Can you please add this? Also why make it a schema instead of a type definition?

Angular is the last of any major framework to support this:

- **Solid.js**: augment `solid-js` module's `JSX.IntrinsicElements`
- **Vue**: augment `vue` module's `GlobalComponents`
- **Svelte**: augment the global `svelteHTML.IntrinsicElements` or augment `svelte/elements` module's `SvelteHTMLElements`
- **React**: augment `react` module's `JSX.IntrinsicElements`
- **Preact**: augment `preact` module's `JSX.IntrinsicElements` with the same definition as for react with react compat enabled
- **Angular**: ???

**Comment by trusktr** [comment]
Alrighty folks, I've opened a new issue that is not a duplicate but a simpler alternative:

- https://github.com/angular/angular/issues/58483

Although I don't have the time to work on it, I hope it serves as a starting point for ideation.

I've been writing custom elements for a long time, and am experienced in writing custom element type definitions for all major frameworks except Angular, so I can provide valuable feedback on what works best for custom element authors and users (for example, React is inflexible, does not cover all use cases, while Solid and Vue allow all use cases to be covered).

**Comment by willmartian** [comment]
This is also affecting @bitwarden. We have a design system built with Angular. We want to use it in non-Angular environments, so we are exploring a rewrite with @lit. Losing template type checking is a no-go, so we will have to create Angular wrapper components. Not fun. Weird that two Google projects interop so poorly.

**Comment by eneajaho** [comment]
Minko recently said that they will be working on this soon: https://youtu.be/CMZHUPSmZx4?t=624

**Comment by break-stuff** [comment]
Is that for custom elements in general or just for custom elements built with Angular Elements?

**Comment by sijakret** [comment]
> Is that for custom elements in general or just for custom elements built with Angular Elements?

I would assume (and sincerely hope) that any solution would operate on the native custom element layer and NOT some proprietary angular layer. Otherwise it would kind of defeat the purpose.
The main use case would be to **enable a typesafe usage of _foreign_ custom elements**,.
For example to be able to use a design system implemented as web components in an angular application.
This is an extremely common use case and the major gap that has been around forever.
It has cost our organization endless headaches, discussions and frustration (with angular not web components).

**Comment by break-stuff** [comment]
I agree, but just wanted to call that out.

**Comment by Deku-nattsu** [comment]
I am implementing my own renderer and stumbled upon this use case and it feels limiting to not have this type of control, also i really hope the improvements support opting out of all built-in DOM elements to allow us to define and use only custom components, in order to allow us to build something similar to [React Three Fiber](https://r3f.docs.pmnd.rs/) or integrate [lynxjs](https://lynxjs.org/)

**Comment by p-b-4** [comment]
My issue was closed as a duplicate: [Narrow the scope of the CUSTOM_ELEMENTS_SCHEMA schema #62484](https://github.com/angular/angular/issues/62484)

So I guess I should share the idea here.

I think it would be a good idea to be able to narrow the scope of a schema within the component template. 

For example with a template notation like this:
```html
@schema(CUSTOM_ELEMENTS_SCHEMA) {
<fluent-card class="custom">
    <fluent-progress value="64"></fluent-progress>
    @schema(ANGULAR_DEFAULT_SCHEMA) {
        <app-example>This component should still exist</app-example>
    }
</fluent-card>
}
```

Or otherwise something like this:
```html
<fluent-card class="custom" @@ignore_unknown>
    <fluent-progress value="64" @@ignore_unknown></fluent-progress>
        <app-example>This component should still exist</app-example>
</fluent-card>
```

My reasoning was, that it's difficult to maintain the quality of a component that is using the schema CUSTOM_ELEMENTS_SCHEMA on the entire template, since this would not be visible from a pull request that only has changes in the template (for example leading to the assumption that the element names are valid, if there are no build errors).

I don't need an entire custom schema, my angle was rather, for scenario's where developers are already using CUSTOM_ELEMENTS_SCHEMA in stead of wrapper libraries. In the current form, it's difficult to see why the schema was added as well, and as said, it leads to confusion when editing or reviewing the template.

**Comment by sijakret** [comment]
> My issue was closed as a duplicate: [Narrow the scope of the CUSTOM_ELEMENTS_SCHEMA schema #62484](https://github.com/angular/angular/issues/62484)

welcome to the party!

i don't think its a good idea to litter all templates with annotations like that.
adding additional components to the set of known native html tags is preferrable in my mind.
from what i recall that data already esists. it's just hardcoded.

the use case for most people is likely: a custom design system built using web components should be usable in angular.

tbh i am not sure why i am wasting my time here though. we're closing in on 10 years

**Comment by p-b-4** [comment]
Thank you.
In my mind they are not competing ideas, but I don't want to protest it being marked duplicate, while I just joined the party (I'm experienced with Angular, but not with discussing it in the community).

The way I see it, the existing CUSTOM_ELEMENTS_SCHEMA is a misleading name from a technical perspective, since it mainly offers a way out of a type of validation, rather than offering a schema to validate specific elements. So I would think that the impact of my request would be a bit lower, since it only requires a change where you can define this way out, rather than customizing how to check which elements are allowed.

Of course it could also help if custom schema's could become available for libraries with web components for example. Especially if this means that you don't need an extensive wrapper library that needs to keep up with the Angular updates in a separate pace from the underlying web component library, that is not maintained specifically for Angular. Being able to use the underlying component library directly, via a custom schema, could potentially reduce complications for updates.

So I think that they are two separate and valid ideas. With a different angle, and different use case.

**Comment by break-stuff** [comment]
So, I know this doesn't solve this problem completely, but I created a PR to add an `isCustomElement` property to modules and components (similar to the way Vuejs handles custom elements) to provide more granular validation. I would love to hear feedback on this.

https://github.com/angular/angular/pull/62638

I am working on something else for the custom element validation piece of it, which may come later this summer.

**Comment by pfteter** [comment]
Any updates on this:
https://www.youtube.com/watch?v=CMZHUPSmZx4&t=624s

**Comment by sijakret** [comment]
> So, I know this doesn't solve this problem completely, but I created a PR to add an `isCustomElement` property to modules and components (similar to the way Vuejs handles custom elements) to provide more granular validation. I would love to hear feedback on this.
> 
> [#62638](https://github.com/angular/angular/pull/62638)
> 
> I am working on something else for the custom element validation piece of it, which may come later this summer.

just to be clear. this proposal was also thrown away without a real explanation, alternative or follow-up. correct?

Edit:
What I mean is:
If this is no good. Fine.
But what‘s the plan then?
Several suggestions have been made..

**Comment by Xiphoseer** [comment]
> just to be clear. this proposal was also thrown away without a real explanation, alternative or follow-up. correct?

Not the maintainer, but no. It was rejected with the comment that adding a user-defined callback in modules/components that the _compiler_ would have to call for every unknown element does not scale.

I do think a more declarative approach (e.g. prefix, allowlist) would have better chances, but I don't see what exactly the maintainers are supposed to suggest on that PR when this issue is about not having a design yet.

**Comment by break-stuff** [comment]
> just to be clear. this proposal was also thrown away without a real explanation, alternative or follow-up. correct?

After working with the Angular team, there are a couple of takeaways:

1. The `CUSTOM_ELEMENTS_SCHEMA` is working as expected. It will exclude the custom elements from validation, but still validate the Angular components. If you are not seeing this, please let me know and possibly provide a sandbox recreation so I can test it and review it with the team.
2. There are proposals in the works to provide custom element extensibility, possibly via a TypeScript type. ETA is unknown.

With that being said, I am also working on a solution for this and will hopefully have something in the next few months.

**Comment by JSMike** [comment]
Not exactly adding new Angular schemas, but I've created a PR to enable loading custom elements manifests so `CUSTOM_ELEMENTS_SCHEMA` isn't required. Feedback appreciated :)
